Rando navi tips

Sarge-117 opened the pull request:
Feature to replace Navi's general overworld quest tips with ones that are specific to rando.

Navi's usual quest tips quickly become meaningless in the context of rando, so this feature will:
1) Check if we're rando'd AND if Navi wants to give us a general overworld tip (based on textId)
2) Replace her textbox with a random tip, drawn from randomizer.cpp

Early draft footage (Navi timer accelerated for testing purpose):
https://cdn.discordapp.com/attachments/988962246585634856/1005975819098792086/randoNavi.mp4

#1086 Used for suggestions for hint text

Sarge-117 committed:
First test of Rando-specific Navi tips

First test of this experiment

Sarge-117 committed:
Fix text colours (Navi text is always light blue)

Navi text should always be \x44

Sarge-117 committed:
Comment clarity

Sarge-117 committed:
Set Navi timer back to normal

Sarge-117 committed:
Update z_message_PAL.c

Sarge-117 committed:
Merge branch 'rando-navi-tips' into rando-next

Sarge-117 committed:
Merge pull request #9 from Sarge-117/rando-next

Rando next merge

Sarge-117 committed:
Update tips

Sarge-117 committed:
Update CMakeLists.txt

Sarge-117 committed:
Comments

Sarge-117 commented:
So far this is working with a sample set of 15 Navi hints.  However, it does not currently use @leggettc18's custom message system.  The new system is a bit over my head, so if we want to switch to that, I might need some guidance!

leggettc18 commented on soh/src/code/z_message_PAL.c:
```diff
@@ -1682,6 +1682,12 @@ void Message_OpenText(GlobalContext* globalCtx, u16 textId) {
+        RandoNaviTip(globalCtx);
+        msgCtx->msgLength = font->msgLength = strlen(font->msgBuf);
+    }
+
```
For the custom-message system, you'll want to move this check to be in the function `CustomMessage_RetrieveIfExists` which is in `OTRGlobals.cpp`, and change it to retrieve a message from a custom message table.

leggettc18 commented on soh/src/code/z_message_RandoNaviTips.c:
```diff
@@ -0,0 +1,313 @@
+#include "global.h"
```
This whole file should probably be moved to be adjacent to or inside of randomizer.cpp to use the custom message system.

Sarge-117 replied:
Sorry @leggettc18, I was just testing out the new system and I think I'm on the right track now.  I can successfully go between two random test messages from Navi.  I'm just looking for the right way to store the information instead of making a huge table manually inside of randomizer.cpp.

leggettc18 replied:
What do you mean "Making a huge table"? Part of the appeal of the `CustomMessageManager` is that it gives you a function to create a table, and the table is only as big as the messages you put into it.

leggettc18 replied:
Ah I see you have it there in randomizer.cpp. I'll comment there.

Sarge-117 replied:
> This whole file should probably be moved to be adjacent to or inside of randomizer.cpp to use the custom message system.

Got the strings at the very bottom of randomizer.cpp for now.

Sarge-117 committed:
Test switch to new message system

Testing setting up the table and calling the messages

leggettc18 commented on soh/src/code/z_message_RandoNaviTips.c:
```diff
@@ -0,0 +1,313 @@
+                                     "Maybe the\x05\x41 boss\x05\x44 has it!\x02");
+                break;
+        }
+    }
```
For the custom-message sytem, you can create a message table and register the messages in a function in `randomizer.cpp` or a new adjacent file, and call that function in InitRando(). You can probably use the `CreateScrubMessages` function as an example. Then you can use the `randoNaviTip` values here as the `textID`s and just run `CustomMessageManager::Instance->RetrieveMessage(<your-table-id>, randoNaviTip)`. Set that equal to a value and have this function return it. `CustomMessage_RetrieveIfExists` handles copying the message into the buffer for you.

leggettc18 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4418,6 +4419,15 @@ void CreateScrubMessages() {
+    customMessageManager->CreateMessage(Randomizer::NaviRandoMessageTableID, 0x01,
+                                        { TEXTBOX_TYPE_BLACK, TEXTBOX_POS_BOTTOM, "Test Eng1", "Test Ger", "Test Fre" });
+}
+
```
Yeah this is on the right track. One thing you can do is create an array of CustomMessageMinimal structs (see `CustomMessageManager.h` for that struct) and iterate over it. That way you at least don't have to write out `TEXTBOX_TYPE_BLACK` and `TEXTBOX_POS_BOTTOM` every time. The function that registers GetItem functions does this, if you want to check it out. Someday we'll have a way to import messages from a JSON file, but that's gonna take a while to build out.

Sarge-117 replied:
Got the array set up now and properly iterating through the 15 sample messages 👍

Sarge-117 committed:
Remove old stuff + add test messages

Sarge-117 committed:
Delete z_message_RandoNaviTips.c

Sarge-117 committed:
Add remaining tips

All the sample hints are now working under the new message system

Sarge-117 committed:
Create const "NaviMsgCount"

So the number of messages doesn't have to be changed in 3 places when adding new entries

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

Sarge-117 committed:
Add German translations

Thanks to Menzelmeister on Discord for translating to German

PurpleHato committed:
ADD: BAGUETTE

Sarge-117 committed:
Merge pull request #10 from PurpleHato/randi-french-navi-tips

ADD: BAGUETTE

Sarge-117 committed:
fix question mark

Sarge-117 committed:
Fix typo, fix Master Sword text

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

Sarge-117 committed:
Fix line break

Sarge-117 committed:
spacing

Sarge-117 commented:
French and German translations are finished for the first set of 15 hints.  Thanks to @PurpleHato and Menzelmeister on discord for the translations!

All three languages tested and working now :)

leggettc18 reviewed (approved):
All looks good to me!

Sarge-117 committed:
Spacing

Sarge-117 commented:
Not sure why Jenkins isn't happy with this (the non-windows builds specifically)

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

leggettc18 commented:
Seems like there's some issues happening right now with some includes causing redefinition errors on other platforms. Something about certain combos in the chain of includes can cause this. I'll see if I can pull this and build it on my Linux laptop tonight to try and figure out what's causing it.

Sarge-117 commented:
> Seems like there's some issues happening right now with some includes causing redefinition errors on other platforms. Something about certain combos in the chain of includes can cause this. I'll see if I can pull this and build it on my Linux laptop tonight to try and figure out what's causing it.

Thanks @leggettc18 that'd be great!

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

Sarge-117 committed:
Change int to u8

Jenkins plz

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

leggettc18 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3975,6 +3976,19 @@ void CreateScrubMessages() {
     }
 }
 
+CustomMessageMinimal NaviMessages[];
```
You should either remove this here, or Remove the `CustomMessageMinimal` type declaration part from line 4083 and specify the size of the array up here instead of down there. The two errors you are getting on Linux are: `array size missing in 'NaviMessages'` on this line and `conflicting declaration 'CustomMesasgeMinimal NaviMessages [15]'` on line 4038.

Additionally, you should probably `#define NUM_NAVI_MESSAGES 15` or something somewhere (maybe `randomizerTypes.h`?) and then use that value when declaring the array size. That way if someone in the future wants to add some there's an easier location to find where to increase the size of the array.

Sarge-117 replied:
Thanks @leggettc18!  Yep, that seems to have fixed it.  I had to reorder the code a bit to put the text array above `CreateNaviRandoMessages()` to avoid the redundant declaration (I originally had it that way because I was trying to put the big block of text at the very bottom for aesthetic purposes).  For some reason VS didn't like having it down below without it being defined a second time.

Anyways, Jenkins is happy now, and I verified that all 3 languages are still working.

leggettc18 commented:
I got around to pulling this on Linux and found where it was erroring, turns out it wasn't the redefinition errors. I left a review comment explaining the situation as well as another point of feedback I had.

Sarge-117 committed:
Add #define, fix conflicting delcaration

Also re-ordered randomizer.cpp a bit to have the text array above CreateNaviRandoMessages()

aMannus reviewed (approved):
Have my useless seal of approval. This is nice! Loving the little extra flair we're giving the SoH rando.

briaguya0 reviewed (commented):
overall this is really cool, but we should probably put it behind a cvar (even if it defaults to on)

aMannus commented:
> overall this is really cool, but we should probably put it behind a cvar (even if it defaults to on)

The point was that the default navi hints are practically useless because rando doesn't follow the regular game's structure of setting flags. I'm pretty sure all the random navi hints are only one textbox as well, so it doesn't cost more time compared to normal.

Tl;dr: I don't see why anyone would want to ever disable these.

Not that it's hard to put it behind an option mind you, I just want to prevent ending up with a ton of checkboxes like the enhancement menu.

briaguya0 commented:
> Tl;dr: I don't see why anyone would want to ever disable these.

i get a kick out of "come on be brave let's go to the deku tree!" hints in locations much later in the game (also 64rando keeps them as is, so it's not just an soh oddity)

tangentially related, i feel like i remember a "disable all navi hints" option existing as a PR at one point, am i imagining that?

aMannus commented:
> > Tl;dr: I don't see why anyone would want to ever disable these.
> 
> i get a kick out of "come on be brave let's go to the deku tree!" hints in locations much later in the game
> 
> tangentially related, i feel like i remember a "disable all navi hints" option existing as a PR at one point, am i imagining that?

Alright, I guess that's fair. We can do the same thing as I do with the random rupee names then. CVar is enabled by default, but can be toggled off. https://github.com/HarbourMasters/Shipwright/pull/1243 for reference.

Not sure about a "disable all navi hints" option. I thought it was already a thing before as well, but I can't seem to find anything about it. I personally probably confused it with the no forced navi option.

briaguya0 commented:
> I personally probably confused it with the no forced navi option.

i feel like i remember someone mentioning a full on disable navi thing when i put out the no forced navi PR but i don't see any comments on it (https://github.com/HarbourMasters/Shipwright/pull/453)

Sarge-117 commented:
Sounds good to me!  I will follow #1243 as an example then.

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-navi-tips

Sarge-117 committed:
Add CVar + Cvar check

Defaults to "1"
